### ISSUE TITLE
Update `cirq.RANDOM_STATE_OR_SEED_LIKE` to support `np.random.Generator`

### ISSUE THREAD

**Issue by NoureldinYosri** [quantumlib/Cirq#6531]
**Is your feature request related to a use case or problem? Please describe.**
We have several modules that spwan multiple threads for performance. each of those threads would be running a random operation using a random state or state. when those threads share the same `RandomState` the multithreading degenerate into sequential processing since those threads will be waiting on write operation of the random state.

**Describe the solution you'd like**
Start supproting `np.random.Generator`. This class provides the same API of `np.random.RandomState` in addition to a `spawn` function which can be used to create independent streams of random values. This will help when starting threads e.g.

```py3
   new_random_generators = prng.spwan(number_threads)
  with ThreadPoolExecutor(max_workers=2) as pool:
          # submit job i with prng new_random_generators[i]
```

**Describe alternatives/workarounds you've considered**
Before calling multithreads generate multipleseeds using a `np.random` or `RandomState`. While this seems like what we are doing with `np.random.Generator.spawn`; it's actually different in that the radom seeds and hence the random sequences created will correlate. This means that when running the same operation (e.g. simulation) multiple times in parallel, the results will correlate.

```
state_0 -> state_1 -> state_2 -> ...
  \         \          \         \
   v         v          v         v
   output_0 output_1   output_2  ....


```
**Additional context (e.g. screenshots)**
The cirq random number support is implemented in cirq-core/cirq/value/random_state.py

**What is the urgency from your perspective for this issue? Is it blocking important work?**
P2 - we should do it in the next couple of quarters

**Comment by mhucka** [comment]
There was a past issue involving `np.random` and multiprocessing. It may not be a concern with the present issue, but on the off chance that there are implications for _this_ issue, it seemed worth pointing to it: #3717 ("Can't use cirq.Simulator() in a multiprocessing closure (unable to pickle)")

**Comment by kitsiosvas** [comment]
Is this issue still open @NoureldinYosri ? I would be interested if so

**Comment by NoureldinYosri** [comment]
@kitsiosvas yes, it's still open. I have an old PR for this https://github.com/quantumlib/Cirq/pull/6566 you can take over it and continue this work if you want to.

**Comment by ANUKOOL324** [comment]
is it still open @NoureldinYosri ? i would be interested !

**Comment by matti-sarjala** [comment]
A small scoping point that might help the next PR:

It may be useful to separate three compatibility questions:

1. **Input normalization**
   - `None`
   - `int` seed
   - `np.random.RandomState`
   - `np.random.Generator`

2. **Backward compatibility**
   - existing `RandomState` behavior should remain unchanged,
   - existing deterministic tests using integer seeds should still produce the same expected behavior unless intentionally updated.

3. **Parallel-safe behavior**
   - new code paths can prefer `np.random.Generator`,
   - threaded/multiprocessing use cases should be able to create independent streams explicitly instead of sharing one mutable random state.

A minimal first acceptance test could cover only the public normalization boundary:

- passing an `int` still works,
- passing `RandomState` still works,
- passing `Generator` now works,
- invalid seed-like objects still raise a clear error.

Then later PRs could migrate internal call sites to prefer `Generator` where parallelism matters. That would avoid turning the first PR into a broad random-number refactor.

**Comment by Prashik123** [comment]
Hi @NoureldinYosri and @mhucka! I would love to pick this up if it is still available.

I have reviewed the previous work in PR #6566 and read through the recent suggestions. I completely agree with @matti-sarjala's proposed scoping. My plan is to tackle this iteratively to ensure stability:

**Phase 1 (This PR):** Focus strictly on the public normalization boundary and backward compatibility. 
* Update `cirq.RANDOM_STATE_OR_SEED_LIKE` (and the relevant parsing functions in `cirq.value.random_state`) to accept `np.random.Generator`.
* Ensure passing an `int` or `np.random.RandomState` remains fully backward compatible and doesn't break existing deterministic tests.
* Add minimal acceptance tests for the new `Generator` input.

Once this foundation is merged, we can handle migrating internal call sites to prefer `Generator` for parallel-safe behavior in a separate follow-up PR.

Please let me know if I can claim this issue and I will get started on the PR!

**Comment by mhucka** [comment]
@Prashik123 Sorry for the late reply. Thank you for  your interest. I've assigned it to you, finally.

**Comment by Prashik123** [comment]
Hi @NoureldinYosri and @mhucka! 

Just dropping a quick note to let you know that I've opened PR #8184 to address this. 

As we discussed, I kept the scope strictly to Phase 1: 
* Updating `RANDOM_STATE_OR_SEED_LIKE` to accept `np.random.Generator`.
* Ensuring 100% backward compatibility for existing `int`, `RandomState`, and `None` inputs.
* Adding the relevant acceptance tests and tightening the error handling.

Looking forward to your feedback whenever you have the time to review it. Thanks again for the guidance and for assigning this to me! 

Best regards,
Prashik

**Comment by arettig** [comment]
Closing this issue in favor of the more detailed/updated #8374.
